File: app/src/main/java/mydebts/android/app/di/SubcomponentBuilderResolver.java
package mydebts.android.app.di;

import android.app.Activity;
import android.support.v4.app.Fragment;
import android.support.annotation.NonNull;

import mydebts.android.app.MyDebtsApplication;

public class SubcomponentBuilderResolver {

    public static SubcomponentBuilder resolve(@NonNull Fragment fragment) {
        return resolve(fragment.getActivity(), fragment.getClass());
    }

    public static SubcomponentBuilder resolve(@NonNull Activity activity) {
        return resolve(activity, activity.getClass());
    }

    private static SubcomponentBuilder resolve(@NonNull Activity activity, Class keyClass) {
        return getApplication(activity)
                .subcomponentBuilder(keyClass);
    }

    private static MyDebtsApplication getApplication(@NonNull Activity activity) {
        return (MyDebtsApplication)activity.getApplication();
    }
}

File: app/src/main/java/mydebts/android/app/di/SubcomponentBuilderResolver.kt
package mydebts.android.app.di

import android.app.Activity
import android.support.v4.app.Fragment

import mydebts.android.app.MyDebtsApplication

object SubcomponentBuilderResolver {

    fun resolve(fragment: Fragment): SubcomponentBuilder {
        return resolve(fragment.activity, fragment.javaClass)
    }

    fun resolve(activity: Activity): SubcomponentBuilder {
        return resolve(activity, activity.javaClass)
    }

    private fun resolve(activity: Activity, keyClass: Class<*>): SubcomponentBuilder {
        return getApplication(activity)
                .subcomponentBuilder(keyClass)
    }

    private fun getApplication(activity: Activity): MyDebtsApplication {
        return activity.application as MyDebtsApplication
    }
}
